### ISSUE TITLE
Strange Isolate.addErrorListener behavior

### ISSUE THREAD

**Issue by yjbanov** [dart-lang/sdk#30425]
The following programs prints nothing and quits with exit code 0.

```dart
import 'dart:io';
import 'dart:isolate';
import 'dart:async';

Future<Null> main() async {
  Isolate.current.addErrorListener(new RawReceivePort((dynamic pair) {
    stdout.writeln('>>> ${pair}');
    stdout.flush();
  }).sendPort);

  await new Future.delayed(const Duration(seconds: 2));
  throw 'a-ha!';
}
```

I expected to see a message on `stdout`.

Update based on conversation below:

It seem the proper fix is some combination of the following:

- make the VM exit with a non-zero exit code
- better document how `addErrorListener` works, specifically its interaction with `setErrorsFatal`
- have `addErrorListener` and `setErrorsFatal` return `Future` that completes after the call's effect is established, so that no magic timers are necessary

**Comment by zanderso** [comment]
Currently, the addErrorListener callback will never run, and the process exits normally because:
- When an Isolate returns to the message loop after an unhandled exception, from then on it will only process OOB messages, meaning that it will never receive messages on the port passed to `addErrorListener`. See the logic [here](https://github.com/dart-lang/sdk/blob/master/runtime/vm/message_handler.cc#L242).
- When an Isolate has an unhandled exception and someone has attached an error listener, the Isolate's "sticky error" field is cleared. See the logic [here](https://github.com/dart-lang/sdk/blob/master/runtime/vm/isolate.cc#L668).

/cc @floitschG to ask if adding an error listener to `Isolate.current` is something that is supposed to actually work.

**Comment by lrhn** [comment]
Have you tried setting errors as non-fatal?
```dart
Isolate.current.setErrorsFatal(false);  // wait for it to be done before causing the error.
```
Otherwise, your throw will terminate the isolate, preventing the receive port from receiving and handling anything.

**Comment by yjbanov** [comment]
Thanks! After `setErrorsFatal(false)` the error listener port now receives the errors and the isolate does not quit.

However, I still don't understand why the original program quits with exit code 0. If errors are fatal, and the reason for quitting the VM is an error, shouldn't it be a non-zero exit code?

**Comment by lrhn** [comment]
I agree that the exit should probably act as if it is caused by an error if the main isolate dies due to an uncaught error.

**Comment by yjbanov** [comment]
The `setErrorsFatal` does not seem to work 100%. The following program fails to continue running the periodic timer after the error.

```dart
import 'dart:async';
import 'dart:isolate';

main() async {
  Isolate.current.setErrorsFatal(false);
  Isolate.current.addErrorListener(new RawReceivePort((event) {
    print('>>> on error: $event');
  }).sendPort);
  Isolate.current.addOnExitListener(new RawReceivePort((event) {
    print('>>> on exit: $event');
  }).sendPort);

  new Timer.periodic(const Duration(seconds: 1), (_) {
    print('>>> tick');
  });

  await new Future.delayed(const Duration(seconds: 3));
  print('>>> throwing');
  throw 'a-ha!';
}
```

**Comment by lrhn** [comment]
Sounds like a bug to me. 
There have been cases of uncaught errors messing up the VM's timer queue, but I thought it had been fixed.

**Comment by rmacnak-google** [comment]
A smaller example where an unhandled exception in one timer causes another timer to be dropped.

```
import 'dart:async';
import 'dart:isolate';
import 'dart:io';

main() async {
  Isolate.current.setErrorsFatal(false);

  new Timer(const Duration(milliseconds: 2), () {
    print("Throwing");
    throw "Something bad";
  });

  new Timer.periodic(const Duration(milliseconds: 3), (_) {
    print('>>> tick');
  });

  sleep(const Duration(milliseconds: 10));
  // Both timers are now due.
}
```

**Comment by rmacnak-google** [comment]
60a7160273f881304a03875cb9c16e8d7b791ab9

**Comment by yjbanov** [comment]
Thanks for the fix! Which version of the SDK will the fix become available? I'd like to test it.

**Comment by zanderso** [comment]
@yjbanov The fix will be in 1.25.0-dev.11.0.

@rmacnak-google I think we should also cherrypick this fix into stable.
